File: backend/app/models/geo.py
"""Models for GEO (Generative Engine Optimization): cross-LLM visibility, prompt
analytics, citation/source mapping, sentiment profiling, and AI readiness audits."""
from __future__ import annotations

from datetime import date
from enum import Enum

from pydantic import Field

from app.models.common import AppModel, Issue


class AnswerEngine(str, Enum):
    """Generative answer engines tracked for brand visibility."""

    CHATGPT = "chatgpt"
    CLAUDE = "claude"
    PERPLEXITY = "perplexity"
    COPILOT = "copilot"
    GEMINI = "gemini"
    GOOGLE_AI_OVERVIEWS = "google_ai_overviews"


class FunnelStage(str, Enum):
    """Marketing funnel stage of a tracked conversational prompt."""

    TOFU = "tofu"  # top of funnel: discovery / education
    MOFU = "mofu"  # middle of funnel: comparison / evaluation
    BOFU = "bofu"  # bottom of funnel: purchase-intent


class SentimentLabel(str, Enum):
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"


class DomainCategory(str, Enum):
    """Classification of a cited source domain for trust analysis."""

    COMMUNITY = "community"          # reddit.com and friends
    ENCYCLOPEDIA = "encyclopedia"    # wikipedia.org
    REVIEW_PLATFORM = "review_platform"  # g2.com, capterra.com, trustpilot.com
    QA_FORUM = "qa_forum"            # quora.com, stackexchange
    NEWS = "news"
    BLOG = "blog"
    OWN_DOMAIN = "own_domain"
    OTHER = "other"


# ---------------------------------------------------------------------------
# Engine answers (provider output)
# ---------------------------------------------------------------------------


class Citation(AppModel):
    """A single URL cited by an answer engine."""

    url: str
    domain: str = ""
    title: str = ""


class EngineAnswer(AppModel):
    """One synthesized answer from one engine for one prompt."""

    engine: AnswerEngine
    prompt: str
    text: str = ""
    ranked_list: list[str] = []  # brand/competitor names in recommendation order
    citations: list[Citation] = []


# ---------------------------------------------------------------------------
# Cross-LLM visibility
# ---------------------------------------------------------------------------


class VisibilityScanRequest(AppModel):
    brand: str
    competitors: list[str] = []
    prompts: list[str] = []  # empty -> use the org's tracked prompt library
    engines: list[AnswerEngine] = []  # empty -> all engines


class VisibilityMetrics(AppModel):
    """Brand visibility aggregates for one engine (or overall)."""

    prompts_scanned: int = 0
    mentions: int = 0
    mention_frequency: float = 0.0  # share of prompts where the brand is mentioned
    share_of_voice: float = 0.0     # brand mentions / (brand + competitor mentions)
    avg_position: float | None = None  # mean rank in generated lists (1 = best)
    mention_frequency_delta: float | None = None
    share_of_voice_delta: float | None = None
    avg_position_delta: float | None = None  # positive = improvement (closer to #1)


class EngineVisibility(VisibilityMetrics):
    engine: AnswerEngine


class VisibilityReport(AppModel):
    brand: str
    competitors: list[str] = []
    scanned_on: date
    engines: list[EngineVisibility] = []
    overall: VisibilityMetrics
    competitor_share_of_voice: dict[str, float] = {}
    answers: list[EngineAnswer] = []


# ---------------------------------------------------------------------------
# Prompt analytics & library
# ---------------------------------------------------------------------------


class TrackedPrompt(AppModel):
    id: str
    text: str
    funnel_stage: FunnelStage = FunnelStage.TOFU
    intent_tags: list[str] = []
    volume_estimate: int = 0
    created_on: date


class PromptCreate(AppModel):
    text: str
    funnel_stage: FunnelStage | None = None  # None -> classified deterministically
    intent_tags: list[str] = []
    volume_estimate: int | None = None       # None -> estimated deterministically


class PromptSuggestion(AppModel):
    text: str
    funnel_stage: FunnelStage = FunnelStage.TOFU
    intent_tags: list[str] = []
    volume_estimate: int = 0


class PromptResearchRequest(AppModel):
    topic: str
    limit: int = Field(default=20, ge=1, le=100)


class PromptResearchResult(AppModel):
    topic: str
    suggestions: list[PromptSuggestion] = []
    total: int = 0


class PromptRankPoint(AppModel):
    day: date
    rank: int | None = None  # None = brand absent from the engine's ranked list


class PromptEngineRank(AppModel):
    engine: AnswerEngine
    current_rank: int | None = None
    previous_rank: int | None = None
    best_rank: int | None = None
    history: list[PromptRankPoint] = []

    @property
    def delta(self) -> int | None:
        if self.current_rank is None or self.previous_rank is None:
            return None
        # Positive delta = improvement (moved closer to #1)
        return self.previous_rank - self.current_rank


class PromptTrackerEntry(AppModel):
    prompt: TrackedPrompt
    engines: list[PromptEngineRank] = []


class PromptTrackerReport(AppModel):
    brand: str
    total_prompts: int = 0
    engines: list[AnswerEngine] = []
    entries: list[PromptTrackerEntry] = []


# ---------------------------------------------------------------------------
# Citation & source mapping
# ---------------------------------------------------------------------------


class CitationScanRequest(AppModel):
    brand: str
    prompts: list[str] = []  # empty -> use the org's tracked prompt library
    engines: list[AnswerEngine] = []
    own_domain: str | None = None  # None -> derived from brand name


class EngineCitations(AppModel):
    engine: AnswerEngine
    citations: list[Citation] = []


class DomainStats(AppModel):
    """Aggregated citation stats + trust analysis for one source domain."""

    domain: str
    category: DomainCategory = DomainCategory.OTHER
    trust_weight: float = Field(default=0.0, ge=0.0, le=1.0)
    citations: int = 0
    frequency: float = 0.0  # share of all citations pointing at this domain
    engines: list[AnswerEngine] = []
    priority_score: float = 0.0  # trust_weight x frequency, 0-100


class CitationReport(AppModel):
    brand: str
    scanned_on: date
    total_citations: int = 0
    engines: list[EngineCitations] = []
    domains: list[DomainStats] = []  # sorted by priority_score desc


class DomainTrustReport(AppModel):
    """Org-wide aggregated view of every domain seen across citation scans."""

    total_citations: int = 0
    domains: list[DomainStats] = []


# ---------------------------------------------------------------------------
# Sentiment profiling
# ---------------------------------------------------------------------------


class SentimentScanRequest(AppModel):
    brand: str
    prompts: list[str] = []  # empty -> use the org's tracked prompt library
    engines: list[AnswerEngine] = []


class SentimentCell(AppModel):
    """One heatmap cell: sentiment of one engine's answer to one prompt."""

    engine: AnswerEngine
    prompt: str
    label: SentimentLabel = SentimentLabel.NEUTRAL
    score: float = Field(default=0.0, ge=-1.0, le=1.0)


class EngineSentiment(AppModel):
    engine: AnswerEngine
    avg_score: float = 0.0
    label: SentimentLabel = SentimentLabel.NEUTRAL


class SentimentReport(AppModel):
    brand: str
    scanned_on: date
    cells: list[SentimentCell] = []       # engine x prompt heatmap
    per_engine: list[EngineSentiment] = []
    overall_score: float = 0.0
    overall_label: SentimentLabel = SentimentLabel.NEUTRAL


# ---------------------------------------------------------------------------
# AI readiness audit
# ---------------------------------------------------------------------------


class SiteArtifacts(AppModel):
    """Raw site artifacts an AI-readiness audit inspects (fetched or injected)."""

    base_url: str
    llms_txt: str | None = None
    robots_txt: str | None = None
    html: str = ""


class ReadinessRequest(AppModel):
    url: str


class AiCrawlerAccess(AppModel):
    crawler: str
    allowed: bool = True


class ReadinessCheck(AppModel):
    code: str
    label: str
    passed: bool
    weight: int = 1
    message: str = ""


class ReadinessReport(AppModel):
    url: str
    score: int = 0  # 0-100 weighted
    grade: str = "F"
    passed_count: int = 0
    total_count: int = 0
    checks: list[ReadinessCheck] = []
    crawler_access: list[AiCrawlerAccess] = []
    fixes: list[Issue] = []  # prioritized: most severe first

File: mcp-server/src/tools/geo.ts
/**
 * GEO (Generative Engine Optimization) MCP tools.
 *
 * Cross-LLM brand visibility scans, conversational prompt research/tracking,
 * citation & source-domain trust mapping, sentiment profiling, and AI
 * readiness audits — backed by the backend's /api/v1/geo/* routes.
 */

import { z } from "zod";
import { apiClient } from "../apiClient.js";
import { buildQuery, defineTool, type ToolDefinition } from "./define.js";

/** Generative answer engines the backend can scan (models/geo.py AnswerEngine). */
const ANSWER_ENGINES = [
  "chatgpt",
  "claude",
  "perplexity",
  "copilot",
  "gemini",
  "google_ai_overviews",
] as const;

const engineEnum = z.enum(ANSWER_ENGINES);

const enginesJsonSchema = {
  type: "array",
  description: "Answer engines to scan (empty/omitted = all engines)",
  items: { type: "string", enum: [...ANSWER_ENGINES] },
} as const;

const promptsJsonSchema = {
  type: "array",
  description: "Prompts to scan (empty/omitted = the org's tracked prompt library)",
  items: { type: "string" },
} as const;

// ---------------------------------------------------------------------------
// geo_visibility_scan — POST /geo/visibility/scan
// ---------------------------------------------------------------------------

const geoVisibilityScanSchema = z.object({
  brand: z.string().min(1, "brand is required"),
  competitors: z.array(z.string().min(1)).optional(),
  prompts: z.array(z.string().min(1)).optional(),
  engines: z.array(engineEnum).optional(),
});

const geoVisibilityScanTool = defineTool(
  "geo_visibility_scan",
  "Scan generative answer engines (ChatGPT, Claude, Perplexity, Copilot, Gemini, Google " +
    "AI Overviews) for brand visibility: mention frequency, share of voice vs competitors, " +
    "and average position in generated recommendation lists, per engine and overall.",
  geoVisibilityScanSchema,
  {
    type: "object",
    properties: {
      brand: { type: "string", description: "Brand name to measure visibility for" },
      competitors: {
        type: "array",
        description: "Competitor brand names for share-of-voice comparison",
        items: { type: "string" },
      },
      prompts: promptsJsonSchema,
      engines: enginesJsonSchema,
    },
    required: ["brand"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/geo/visibility/scan", args)
);

// ---------------------------------------------------------------------------
// geo_prompt_research — POST /geo/prompts/research
// ---------------------------------------------------------------------------

const geoPromptResearchSchema = z.object({
  topic: z.string().min(1, "topic is required"),
  limit: z.number().int().min(1).max(100).optional(),
});

const geoPromptResearchTool = defineTool(
  "geo_prompt_research",
  "Research conversational prompts people ask AI assistants about a topic, classified by " +
    "funnel stage (tofu/mofu/bofu) with intent tags and volume estimates — candidates for " +
    "the org's tracked prompt library.",
  geoPromptResearchSchema,
  {
    type: "object",
    properties: {
      topic: { type: "string", description: "Topic or niche to research prompts for" },
      limit: { type: "integer", minimum: 1, maximum: 100, description: "Maximum suggestions to return (default 20)" },
    },
    required: ["topic"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/geo/prompts/research", args)
);

// ---------------------------------------------------------------------------
// geo_prompt_tracker — GET /geo/prompts/tracker
// ---------------------------------------------------------------------------

const geoPromptTrackerSchema = z.object({
  brand: z.string().min(1, "brand is required"),
  competitors: z.array(z.string().min(1)).optional(),
  engines: z.array(engineEnum).optional(),
});

const geoPromptTrackerTool = defineTool(
  "geo_prompt_tracker",
  "Track the brand's rank over time in each answer engine's generated recommendation list " +
    "for every tracked prompt: current/previous/best rank plus a per-day rank history.",
  geoPromptTrackerSchema,
  {
    type: "object",
    properties: {
      brand: { type: "string", description: "Brand name to track ranks for" },
      competitors: {
        type: "array",
        description: "Competitor brand names to include in ranked lists",
        items: { type: "string" },
      },
      engines: enginesJsonSchema,
    },
    required: ["brand"],
    additionalProperties: false,
  },
  (args) =>
    apiClient.get(
      `/geo/prompts/tracker${buildQuery({
        brand: args.brand,
        competitors: args.competitors,
        engines: args.engines,
      })}`
    )
);

// ---------------------------------------------------------------------------
// geo_citations_scan — POST /geo/citations/scan
// ---------------------------------------------------------------------------

const geoCitationsScanSchema = z.object({
  brand: z.string().min(1, "brand is required"),
  prompts: z.array(z.string().min(1)).optional(),
  engines: z.array(engineEnum).optional(),
  own_domain: z.string().optional(),
});

const geoCitationsScanTool = defineTool(
  "geo_citations_scan",
  "Scan which URLs/domains answer engines cite when answering brand-relevant prompts, " +
    "aggregated into per-domain stats with category, trust weight, citation frequency, and " +
    "a priority score for outreach.",
  geoCitationsScanSchema,
  {
    type: "object",
    properties: {
      brand: { type: "string", description: "Brand name the prompts relate to" },
      prompts: promptsJsonSchema,
      engines: enginesJsonSchema,
      own_domain: { type: "string", description: "The brand's own domain (omitted = derived from the brand name)" },
    },
    required: ["brand"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/geo/citations/scan", args)
);

// ---------------------------------------------------------------------------
// geo_domain_trust — GET /geo/citations/domains
// ---------------------------------------------------------------------------

const geoDomainTrustSchema = z.object({
  own_domain: z.string().optional(),
});

const geoDomainTrustTool = defineTool(
  "geo_domain_trust",
  "Get the org-wide aggregated trust analysis of every source domain seen across citation " +
    "scans: category (community, encyclopedia, review platform, ...), trust weight, citation " +
    "counts, and priority score, sorted by priority.",
  geoDomainTrustSchema,
  {
    type: "object",
    properties: {
      own_domain: { type: "string", description: "The brand's own domain, to classify it separately" },
    },
    required: [],
    additionalProperties: false,
  },
  (args) => apiClient.get(`/geo/citations/domains${buildQuery({ own_domain: args.own_domain })}`)
);

// ---------------------------------------------------------------------------
// geo_sentiment_scan — POST /geo/sentiment/scan
// ---------------------------------------------------------------------------

const geoSentimentScanSchema = z.object({
  brand: z.string().min(1, "brand is required"),
  prompts: z.array(z.string().min(1)).optional(),
  engines: z.array(engineEnum).optional(),
});

const geoSentimentScanTool = defineTool(
  "geo_sentiment_scan",
  "Profile how answer engines talk about a brand: an engine x prompt sentiment heatmap " +
    "(positive/neutral/negative with -1..1 scores), per-engine averages, and an overall " +
    "sentiment label.",
  geoSentimentScanSchema,
  {
    type: "object",
    properties: {
      brand: { type: "string", description: "Brand name to profile sentiment for" },
      prompts: promptsJsonSchema,
      engines: enginesJsonSchema,
    },
    required: ["brand"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/geo/sentiment/scan", args)
);

// ---------------------------------------------------------------------------
// geo_ai_readiness_audit — POST /geo/readiness/audit
// ---------------------------------------------------------------------------

const geoAiReadinessAuditSchema = z.object({
  url: z.string().min(1, "url is required"),
});

const geoAiReadinessAuditTool = defineTool(
  "geo_ai_readiness_audit",
  "Audit a site's readiness for AI answer engines: llms.txt, robots.txt access for AI " +
    "crawlers (GPTBot, ClaudeBot, PerplexityBot, ...), structured data, and content signals. " +
    "Returns a weighted 0-100 score, grade, per-check results, and prioritized fixes.",
  geoAiReadinessAuditSchema,
  {
    type: "object",
    properties: {
      url: { type: "string", description: "Absolute URL of the site to audit, e.g. https://example.com" },
    },
    required: ["url"],
    additionalProperties: false,
  },
  (args) => apiClient.post("/geo/readiness/audit", args)
);

// ---------------------------------------------------------------------------

/** All GEO tools exposed by this MCP server. */
export const geoTools: ToolDefinition[] = [
  geoVisibilityScanTool,
  geoPromptResearchTool,
  geoPromptTrackerTool,
  geoCitationsScanTool,
  geoDomainTrustTool,
  geoSentimentScanTool,
  geoAiReadinessAuditTool,
];
